xonsh/history/json.py

```diff
@@ -0,0 +1,459 @@
+            return [self[i] for i in range(*key.indices(size))]
+        elif not isinstance(key, int):
+            raise IndexError(
+                'JsonCommandField may only be indexed by int or slice.')
+        elif size == 0:
+            raise IndexError('JsonCommandField is empty.')
+        # now we know we have an int
+        key = size + key if key < 0 else key  # ensure key is non-negative
+        bufsize = len(self.hist.buffer)
+        if size - bufsize <= key:  # key is in buffer
+            return self.hist.buffer[key + bufsize - size].get(
+                self.field, self.default)
+        # now we know we have to go into the file
+        queue = self.hist._queue
+        queue.append(self)
+        with self.hist._cond:
+            self.hist._cond.wait_for(self.i_am_at_the_front)
+            with open(self.hist.filename, 'r', newline='\n') as f:
+                lj = xlj.LazyJSON(f, reopen=False)
+                rtn = lj['cmds'][key].get(self.field, self.default)
+                if isinstance(rtn, xlj.LJNode):
+                    rtn = rtn.load()
+            queue.popleft()
+        return rtn
+
+    def i_am_at_the_front(self):
+        """Tests if the command field is at the front of the queue."""
+        return self is self.hist._queue[0]
+
+
+# Interface to History
+class JsonHistory(HistoryBase):
+    """Xonsh session history.
+
+    Indexing
+    --------
+    History object acts like a sequence that can be indexed in a special way
+    that adds extra functionality. At the moment only history from the
+    current session can be retrieved. Note that the most recent command
+    is the last item in history.
+
+    The index acts as a filter with two parts, command and argument,
+    separated by comma. Based on the type of each part different
+    filtering can be achieved,
+
+        for the command part:
+
+            - an int returns the command in that position.
+            - a slice returns a list of commands.
+            - a string returns the most recent command containing the string.
+
+        for the argument part:
+
+            - an int returns the argument of the command in that position.
+            - a slice returns a part of the command based on the argument
+              position.
+
+    The argument part of the filter can be omitted but the command part is
+    required.
+
+    Command arguments are separated by white space.
+
+    If the filtering produces only one result it is
+    returned as a string else a list of strings is returned.
+
+    Attributes
+    ----------
+    rtns : sequence of ints
+        The return of the command (ie, 0 on success)
+    inps : sequence of strings
+        The command as typed by the user, including newlines
+    tss : sequence of two-tuples of floats
+        The timestamps of when the command started and finished, including
+        fractions
+    outs : sequence of strings
+        The output of the command, if xonsh is configured to save it
```

Comment:
Note that these attrs are part of the base interface, but they only seem to be present on the JSON history, unless I am missing something....